### ISSUE TITLE
Pyright Fails without variable assignment with SQLAlchemy Select type

### ISSUE THREAD

**Issue by ManiMozaffar** [microsoft/pyright#8120]
The function signature is saying the return type is select of string. but i'm selecting Literal[1]. so type checker must always complain. but it doesn't always! It is very weird, when I try to assign it to variable and return it, type checker complain. When i just return it, type checker doesn't complain.
I tried reproducing it with python's code without sqla dependency but it's hard to actually understand and debug how the typing class hierarchy works in SQLA codebase. I couldn't reproduce it by combination of typevar (bounded to tuple) and generic type.

![image](https://github.com/microsoft/pyright/assets/86913668/d3a7cac4-aa9d-4aa9-81a6-f4bc9a52a93e)

To Reproduce

```python
import sqlalchemy as sa


def correct() -> sa.Select[tuple[str]]:
    q = sa.select(sa.literal(1))
    return q


def wrong() -> sa.Select[tuple[str]]:
    return sa.select(sa.literal(1))
```

**Comment by erictraut** [comment]
Pyright's behavior here is correct, so I don't consider this a bug.

In your `wrong` example, pyright is using bidirectional type inference. That means it uses the "expected type" to prepopulate type constraints. In this specific example, the expected type (which comes from the return type annotation) is `Select[tuple[str]]`. Then, when pyright performs overload matching, it attempts to find an overload that works with the provided arguments. There are 11 overloaded signatures provided for the `select` function. Of these, 9 can immediately be eliminated as candidates because they have the wrong arity (i.e. the wrong number of parameters). the remaining two are:

```python
@overload
def select(__ent0: _TCCA[_T0]) -> Select[Tuple[_T0]]: ...
@overload
def select(
    *entities: _ColumnsClauseArgument[Any], **__kw: Any
) -> Select[Any]: ...
```

Given that `_T0` has an upper bound of `str` (due to the constraint that was imposed by the expected type), the first overload is not valid. However, the second of these two overloads is valid. It has a return type of `Select[Any]` which is compatible with the return type `Select[Tuple[str]]`, so there is no type error.

In your first example (`correct`), there is no inference context (i.e. no "expected type"), so no additional constraints are present. In this case, the first overload matches, and the expression `sa.select(sa.literal(1))` evaluates to `Select[tuple[int]]`. The variable `q` takes on the type `Select[tuple[int]]` by assignment. In the next statement, this type is not compatible with the return type `Select[tuple[str]]`, hence the type error.

**Comment by ManiMozaffar** [comment]
Thanks for the answer, it makes sense!

**Comment by CaselIT** [comment]
@erictraut I suppose there is no way currently of having the type overload do the correct thing, right?

IIRC the last time I looked the missing thing is that unpack cannot be nested into a generic type, so something like this that would replace all the overload, does not work
```py
select(*entities: _ColumnsClauseArgument[Unpack[T]]) -> Select[Unpack[T]]
```
